Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -350,41 +360,58 @@ private class BetterHTTPParser {
+
             if let bytes = bytes {
                 self.rawBytesView = bytes
                 defer {
                     self.rawBytesView = .init(start: UnsafeRawPointer(bitPattern: 0xdafbabe), count: 0)
                 }
-                parserResult = c_nio_http_parser_execute_swift(parserPtr,
-                                                               &self.settings,
-                                                               bytes.baseAddress! + self.httpParserOffset,
-                                                               bytes.count - self.httpParserOffset)
+
+                let startPointer = bytes.baseAddress! + self.httpParserOffset
+                let bytesToRead = bytes.count - self.httpParserOffset
+
+                rc = c_nio_llhttp_execute(parserPtr,
+                                          startPointer,
+                                          bytesToRead)
+
+                if rc == HPE_PAUSED_UPGRADE {
+                    // This is a special pause. We don't need to stop here (our other code will prevent us
+                    // parsing past this point, but we do need a special hook to work out how many bytes were read.
+                    // The force-unwrap is safe: we know we hit an "error".
+                    bytesRead = UnsafeRawPointer(c_nio_llhttp_get_error_pos(parserPtr)!) - startPointer
+                    c_nio_llhttp_resume_after_upgrade(parserPtr)
+                    rc = HPE_OK
+                } else {
+                    bytesRead = bytesToRead
+                }
             } else {
-                parserResult = c_nio_http_parser_execute(parserPtr, &self.settings, nil, 0)
+                rc = c_nio_llhttp_finish(parserPtr)
+                bytesRead = 0
             }
-            parserErrno = parserPtr.pointee.http_errno
-            return parserResult
+
+            return rc
         }
-        assert(parserConsumed >= 0)
+
         // self.parser must be non-nil here because we can't be re-entered here (ByteToMessageDecoder guarantee)
-        guard parserErrno == 0 else {
+        guard parserErrno == HPE_OK else {
             // if we chose to abort (eg. wrong HTTP version) the error will be in self.httpErrno, otherwise http_parser
             // will tell us...
             // self.parser must be non-nil here because we can't be re-entered here (ByteToMessageDecoder guarantee)
-            // If we have a richer error than the errno code, and the errno is unknown, we'll use it. Otherwise, we use the
+            // If we have a richer error than the errno code, and the errno is internal, we'll use it. Otherwise, we use the
             // error from http_parser.
-            let err = self.httpErrno ?? http_errno(rawValue: parserErrno)
-            if err == HPE_UNKNOWN, let richerError = self.richerError {
+            let err = self.httpErrno ?? parserErrno
+            if (err == HPE_INTERNAL || err == HPE_USER), let richerError = self.richerError {
```

Comment:
Why also `HPE_USER` here?

Reply:
Two different codepaths. Annoyingly, we get HPE_INTERNAL from one, and HPE_USER from the other.